sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala

```diff
@@ -784,121 +853,446 @@ class SumBinaryFixer extends BatchedRunningWindowFixer with Arm with Logging {
+            // So if the group by mask in not a constant, then the order by mask also cannot be
+            // a constant
+            throw new IllegalStateException(
+              "Internal Error the order mask is not a subset of the part mask")
+        }
+    }
+
+    logDebug(s"dense rank: updateState from $previousRank to...")
+    previousRank.foreach(_.close)
+    previousRank = Some(ret.getScalarElement(ret.getRowCount.toInt - 1))
+    logDebug(s"dense rank: ... $previousRank")
+
+    ret
+  }
+
+  override def close(): Unit = {
+    previousRank.foreach(_.close())
+  }
+}
+
+object DenseRankFixer extends Arm {
+  private[this] def isFirstTrue(cv: cudf.ColumnView): Boolean = {
+    withResource(cv.getScalarElement(0)) { scalar =>
+      scalar.getBoolean
+    }
+  }
+
+  private def addPrevRank(cv: cudf.ColumnView, prevRank: Scalar): cudf.ColumnVector =
+    cv.add(prevRank)
+
+  private[this] def addPrevRankMinusOne(cv: cudf.ColumnView,
+      prevRank: Scalar): cudf.ColumnVector = {
+    withResource(addPrevRank(cv, prevRank)) { prev =>
+      withResource(Scalar.fromInt(1)) { one =>
+        prev.sub(one)
+      }
+    }
+  }
+
+  private def fixRankInSamePartition(
+      rank: cudf.ColumnView,
+      orderMask: cudf.ColumnView,
+      prevRank: Scalar): cudf.ColumnVector = {
+    // This is a little ugly, but the only way to tell if we are part of the previous order by
+    // group or not is to look at the orderMask. In this case we check if the first value in the
+    // mask is true.
+    val added = if (isFirstTrue(orderMask)) {
+      addPrevRankMinusOne(rank, prevRank)
+    } else {
+      addPrevRank(rank, prevRank)
+    }
+    withResource(added) { added =>
+      orderMask.ifElse(prevRank, added)
+    }
+  }
+}
+
+/**
+ * Rank is a special window operation where it is only supported as a running window. In cudf
+ * it is only supported as a scan and a group by scan. But there are special requirements beyond
+ * that when doing the computation as a running batch. To fix up each batch it needs both the rank
+ * and the row number. To make this work and be efficient there is different behavior for batched
+ * running window vs non-batched. If it is for a running batch we include the row number values,
+ * in both the initial projections and in the corresponding aggregations. Then we combine them
+ * into a struct column in `scanCombine` before it is passed on to the `RankFixer`. If it is not
+ * a running batch, then we drop the row number part because it is just not needed.
+ * @param children the order by columns.
+ */
+case class GpuRank(children: Seq[Expression]) extends GpuRunningWindowFunction
+    with GpuBatchedRunningWindowWithFixer {
   override def nullable: Boolean = false
   override def dataType: DataType = IntegerType
 
-  override def children: Seq[Expression] = Nil
+  // RUNNING WINDOW ONLY - RANK DOES NOT WORK FOR ANYTHING ELSE IN SPARK
```

Comment:
Didn't expect an all-caps comment here which looks almost like a TODO.  Should this be a `@note` in the class description?

This applies to a couple of other all-caps comments below as well.